src/components/feedback/Dialog.tsx

```diff
@@ -0,0 +1,158 @@
+import classnames from 'classnames';
+import type { RefObject } from 'preact';
+import { useEffect, useLayoutEffect } from 'preact/hooks';
+
+import { useElementShouldClose } from '../../hooks/use-element-should-close';
+import { useSyncedRef } from '../../hooks/use-synced-ref';
+import { useUniqueId } from '../../hooks/use-unique-id';
+import type { PresentationalProps } from '../../types';
+import { downcastRef } from '../../util/typing';
+import Overlay from '../layout/Overlay';
+import Panel from '../layout/Panel';
+import type { PanelProps } from '../layout/Panel';
+
+type ComponentProps = {
+  /**
+   * Dialog _should_ be provided with a close handler. We have a few edge use
+   * cases, however, in which we need to render a "non-closeable" modal.
+   */
+  onClose?: () => void;
+  width?: 'sm' | 'md' | 'lg' | 'custom';
+  /**
+   * Element that should take focus when the Dialog is first rendered. When not
+   * provided ("auto"), the dialog's outer element will take focus. Setting this
+   * prop to "manual" will opt out of focus routing.
+   */
+  initialFocus?: RefObject<HTMLOrSVGElement | null> | 'auto' | 'manual';
+};
+
+// This component forwards a number of props on to `Panel` but always sets the
+// `fullWidthHeader` prop to `true`.
+export type DialogProps = PresentationalProps &
+  ComponentProps &
+  Omit<PanelProps, 'fullWidthHeader'>;
+
+const noop = () => {};
+
+/**
+ * Show a modal
+ */
+const DialogNext = function Dialog({
+  children,
+  width = 'md',
+
+  classes,
+  elementRef,
+  initialFocus = 'auto',
+
+  // Forwarded to Panel
+  buttons,
+  icon,
+  onClose,
+  paddingSize = 'md',
+  scrollable = true,
+  title,
+
+  ...htmlAttributes
+}: DialogProps) {
+  const modalRef = useSyncedRef(elementRef);
+  const closeHandler = onClose ?? noop;
+  useElementShouldClose(modalRef, true, closeHandler);
+
+  const dialogDescriptionId = useUniqueId('dialog-description');
+
+  useEffect(() => {
+    if (initialFocus === 'manual') {
+      return;
+    }
+    if (initialFocus === 'auto') {
+      // An explicit `initialFocus` has not be set, so use automatic focus
+      // handling. Modern accessibility guidance is to focus the dialog itself
+      // rather than trying to be smart about focusing a particular control
+      // within the dialog.
+      modalRef.current?.focus();
+      return;
+    }
+
+    const focusEl = initialFocus?.current as HTMLElement & {
+      disabled?: boolean;
+    };
+
+    if (focusEl && !focusEl.disabled) {
+      focusEl.focus();
+    } else {
+      // Fall back to focusing the modal itself
+      modalRef.current?.focus();
+      return;
+    }
```

Comment:
I think you can disregard this `else` block entirely, as you have already checked for `'manual'` and `'auto'` values before getting to this point.

Reply:
I'm going to change the accepted values for `initialFocus` very soon, and will clean this up at that time!